scheduler/reconcile.go

```diff
@@ -680,58 +590,66 @@ func (a *allocReconciler) handleGroupCanaries(all allocSet, desiredChanges *stru
 	return canaries, all
 }
 
-// computeLimit returns the placement limit for a particular group. The inputs
-// are the group definition, the untainted, destructive, and migrate allocation
-// set and whether we are in a canary state.
-func (a *allocReconciler) computeLimit(group *structs.TaskGroup, untainted, destructive, migrate allocSet, canaryState bool) int {
+// computeUnderProvisionedBy returns the number of allocs that still need to be
+// placed for a particular group. The inputs are the group definition, the untainted,
+// destructive, and migrate allocation sets, and whether we are in a canary state.
+// The following rules are applied:
+// * If no update strategy, or nothing is migrating, being replaced, or reconnecting,
+//   allow as many as defined in group.Count
+// * If the deployment is paused, failed, or canarying, allow no placements
+// * If the deployment is nil, allow MaxParallel placements
+// * If the deployment is not nil, but has unhealthy allocs, allow no placements
+// * If the deployment is not nil, and all allocs are healthy, allow MaxParallel
+//   minus the number of healthy allocs placements
+// * If the calculated limit is less than zero, return 0
```

Comment:
I went out of my way to not do much cleaning up of the internal code, but that said, I did pause before writing this because I know how easily comments can get out of sync with the code. How do you feel about moving these into the body of the function to improve the comments that are there already?

Reply:
I took a stab at merging these comments with the extant inline comments to just make the comments a little better. I also refactored the code a bit.